### PR TITLE
Close PSBT leaks found by Christian, fixes to detect them in future

### PR DESCRIPTION
**Pull request by rustyrussell**
Interfacing with libwally without wrapping everything in our own struct is a pain, but this now manages it.

As an aside, there are cleanups and clarifications to the memleak interface.

### CODE REVIEW

File: bitcoin/psbt.c [hunk 1]
```diff
@@ -32,8 +32,12 @@ static struct wally_psbt *init_psbt(const tal_t *ctx, size_t num_inputs, size_t
 	else
 		wally_err = wally_psbt_init_alloc(0, num_inputs, num_outputs, 0, &psbt);
 	assert(wally_err == WALLY_OK);
+	tal_steal(ctx, psbt);
+	/* If both of these are zero, no sub-allocations were done */
+	if (num_inputs || num_outputs)
```
Comment (cdecker) [inline bitcoin/psbt.c:37]: Is there any harm in calling it unconditionally?

File: common/daemon.c [hunk 2]
```diff
@@ -81,11 +81,20 @@ void send_backtrace(const char *why)
...
 	if (t)
 		errx(1, "Outstanding taken pointers: %s", t);
 
+	wally_leak = tal_first(wally_tal_ctx);
+	if (wally_leak) {
+		/* Trigger valgrind to tell us about this! */
+		tal_free(wally_leak);
+		*wally_leak = 0;
+		errx(1, "Outstanding wally allocations");
+	}
```
Comment (cdecker) [inline common/daemon.c:96]: Very nice, this will report the location the non-collected child was allocated at in `valgrind`, correct?

**Review by cdecker — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/4071/commits/5677d0af4b15134cf8490595e6e64b4a8785bf62

**Comment by cdecker** [comment]
After checking I can confirm that the memleaks are indeed gone:

memdump-after

However it seems like I produced a (slow) memory leak myself when precomputing the txids in the speedup PR, I'll need to look into dropping the `b->txids` at an opportune time. Afterwards we'll have a 115MB memory footprint, 104 of which is logs :-)

File: common/daemon.c (hunk 2 above)
Comment (rustyrussell) [reply common/daemon.c:96]: Yes, that's how I found them!

File: bitcoin/psbt.c (hunk 1 above)
Comment (rustyrussell) [reply bitcoin/psbt.c:37]: No *harm*, but we have a debug check for the moment that we don't call tal_gather_wally unless there's something to gather.  Found some places where I was being overzealous.

**Pull request merged**